### PR TITLE
NGI-RNAseq

### PR DESCRIPTION
**Pull request by Galithil**
This is the first draft at adding the rna-seq pipeline as an engine of the ngi_pipeline. It requires a couple of changes in the config file as well, and it moves a lot of the logic to the piper engine, as we need this engine to start on a project rather than on individual samples. 

It currently starts the rna_pipeline properly on milou, I'll perform additional tests to ensure that the piper engine was not impacted.

### CODE REVIEW

File: ngi_pipeline/conductor/launchers.py [hunk 1]
```diff
@@ -25,15 +25,15 @@ def launch_analysis(projects_to_analyze, restart_failed_jobs=False,
...
-            LOG.error('Project {} could not be processed: {}'.format(project, e))
-            continue
-        engine.local_process_tracking.update_charon_with_local_jobs_status(config=config)
     charon_session = CharonSession()
     for project in projects_to_analyze:
+        analysis=NGIAnalysis(project, restart_failed_jobs,
+                    restart_finished_jobs, restart_running_jobs,
+                    keep_existing_data, no_qc, exec_mode,
+                    quiet, manual, config, config_file_path,
+                    generate_bqsr_bam, LOG)
```
Comment (b97pla) [inline ngi_pipeline/conductor/launchers.py]: for clarity/robustness, it may be worthwhile to use the names for the arguments or pass them as a `**{}`

File: ngi_pipeline/engines/piper_ngi/launchers.py [hunk 2]
```diff
@@ -432,16 +459,17 @@ def sbatch_piper_sample(command_line_list, workflow_name, project, sample,
     with open(sbatch_outfile, 'w') as f:
         f.write("\n".join(sbatch_text_list))
     LOG.info("Queueing sbatch file {} for job {}".format(sbatch_outfile, job_identifier))
+    slurm_job_id=7777777
```
Comment (b97pla) [inline ngi_pipeline/engines/piper_ngi/launchers.py]: is this code left from debugging?

Comment (Galithil) [inline ngi_pipeline/engines/piper_ngi/launchers.py]: Yes, I'm still in the process of making sure that the piper engine was not impacted.

File: ngi_pipeline/engines/piper_ngi/launchers.py [hunk 3]
```diff
@@ -46,165 +48,190 @@
...
+                    except RuntimeError as e:
+                        LOG.error(e)
+                        ## Question: should we just kill the run in this case or let it go?
+                        continue
+                except (NotImplementedError, RuntimeError, ValueError) as e:
+                    error_msg = ('Processing project "{}" / sample "{}" / workflow "{}" '
+                                 'failed: {}'.format(analysis_object.project, sample,
+                                                     workflow_subtask,
+                                                     e))
+                    LOG.error(error_msg)
```
Comment (b97pla) [inline ngi_pipeline/engines/piper_ngi/launchers.py]: I've always found this code quite hard to follow, so refactoring to analysis objects is probably a good idea that could be expanded upon more. I'm not confident enough in this code section to do a proper code review though, I think @vezzi should take a look at this as well.

File: ngi_pipeline/engines/rna_ngi/__init__.py [hunk 4]
```diff
@@ -0,0 +1,2 @@
+from .launchers import analyze
+from .local_process_tracking import update_charon_with_local_jobs_status
```
Comment (b97pla) [inline ngi_pipeline/engines/rna_ngi/__init__.py]: what does the dot before the module name mean?

File: ngi_pipeline/engines/piper_ngi/launchers.py (hunk 3 above)
Comment (Galithil) [inline ngi_pipeline/engines/piper_ngi/launchers.py]: The main issue that caused me to work on this is that the current code assumes that we want to start analysis on sample level, no matter the engine. Unfortunately, this doesnt hold true for the next two engines, namely RNA and Cancer. as such, I needed to move big chunks of code around, and the obnoxious amount of parameters that have to be carried over different functions is really something I'm not a fan of. 
FV is still on holidays until the end of this week, and I'll leave afterwards, so we might put this on hold.

File: ngi_pipeline/engines/rna_ngi/__init__.py (hunk 4 above)
Comment (Galithil) [inline ngi_pipeline/engines/rna_ngi/__init__.py]: it's the same as ./, it's to ensure that you load the right "launchers" since we have at least three files sharing the same name.

File: ngi_pipeline/engines/rna_ngi/launchers.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+                elif charon_reported_status == "FAILED":
+                    if not analysis_object.restart_failed_jobs:
+                        error_text = ('FAILED:  Project "{}" / sample "{}" Charon reports '
+                                      'FAILURE, manual investigation needed!'.format(analysis_object.project, sample))
+                        LOG.error(error_text)
+                        if not analysis_object.config.get('quiet'):
+                            mail_analysis(project_name=analysis_object.project.name, sample_name=sample.name,
+                                          engine_name=analysis_module.__name__,
+                                          level="ERROR", info_text=error_text)
+                        continue
```
Comment (b97pla) [inline ngi_pipeline/engines/rna_ngi/launchers.py]: There is so much code repetition in these modules and especially the analyze methods are hard to follow. It would really be great if these bits could be refactored to be clearer and to re-use code rather than repeat it.

Comment (Galithil) [inline ngi_pipeline/engines/rna_ngi/launchers.py]: I'm not sure I agree. 
While this behaviour is the same for the two engines, it might require tweaking. For instance, I changed how the libprep QC was handled in the RNA engine, but not in the piper engine. 

If you're talking about refactoring the code intra-engine, however, I don't think it can be refactored much more than what it is. What did you have in mind ?

Comment (b97pla) [inline ngi_pipeline/engines/rna_ngi/launchers.py]: I haven't studied the code deeply, so I can't be very specific but to me, there seems to be a lot of functionality that is common to the engines (checking analysis status, reporting errors, mailing errors, check if jobs should be restarted etc. etc.). Even if the overall logic can be reused, perhaps that kind of helper functionality can?

Comment (Galithil) [inline ngi_pipeline/engines/rna_ngi/launchers.py]: Okay, I moved around the functions handling the status and now both piper and the rna_ngi pipeline use them.

Comment (b97pla) [inline ngi_pipeline/engines/rna_ngi/launchers.py]: To be a bit specific, I had a crack at rewriting lines 25-87 of this function (just intra-engine stuff) [here](https://gist.github.com/b97pla/b2574f551d91e54f1b6be9b0c89302df) (not tested). It's just minor stuff to reduce repetition but IMHO, I think the logic is immediately easier to follow.

**Comment by Galithil** [comment]
Okay, it should be mostly ready. Piper seems to still work properly, as I can start jobs, and the rna should work fine as well. Let's wait for @vezzi if you wish.

**Comment by b97pla** [comment]
Yes, Vezzi needs to sign off on this and he should be satisfied with the integration testing.

File: ngi_pipeline/conductor/classes.py [hunk 6]
```diff
@@ -69,3 +111,45 @@ def add_fastq_files(self, fastq):
...
+    charon_session = CharonSession()
+    try:
+        best_practice_analysis = charon_session.project_get(project.project_id)["best_practice_analysis"]
+        if not best_practice_analysis:
+            raise KeyError("For once in my life ever can't you just fill in the forms properly")
+    except KeyError:
+        error_msg = ('No best practice analysis specified in Charon for '
+                     'project "{}". Using "whole_genome_reseq"'.format(project))
+        LOG.error(error_msg)
+        best_practice_analysis = "whole_genome_reseq"
```
Comment (vezzi) [inline ngi_pipeline/conductor/classes.py]: Would be better to have qc has default engine. Connected to this would be better to have qc has a proper engine

File: ngi_pipeline/engines/utils.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        if not analysis_object.restart_failed_jobs:
+            error_text = ('FAILED:  Project "{}" / sample "{}" Charon reports '
+                          'FAILURE, manual investigation needed!'.format(analysis_object.project, sample))
+            LOG.error(error_text)
+            if not analysis_object.config.get('quiet'):
+                mail_analysis(project_name=analysis_object.project.name, sample_name=sample.name,
+                              engine_name=analysis_module.__name__,
+                              level="ERROR", info_text=error_text)
+            return False 
+    else:
```
Comment (vezzi) [inline ngi_pipeline/engines/utils.py]: let me see if I understand well.
If `charon_reported_status ` is `ANALYZED` but I want to rerun the sample anyway (i.e., `analysis_object.restart_finished_jobs = TRUE`, then line 41 evaluates to `FALSE`. So I will not enter in any other else/if section and this function will return `None`, therefore in piper_ngi/launchers.py inline 66 I will evaluate to `TRUE` and do not restart the analysis....

**Comment by ewels** [comment]
@vezzi @b97pla - ok to merge now?

**Comment by vezzi** [comment]
well I think no, as if we merge now into master this will be the version we will deploy in Irma next time we have to deploy ngi_pipeline. So we can merge it into a devel branch but not into the master until we cannot properly test it

**Comment by ewels** [comment]
Ok cool. Looks like GitHub [recently added the ability to switch branch on a PR](https://github.com/blog/2224-change-the-base-branch-of-a-pull-request) so should be easy to do.

**Comment by johanherman** [comment]
@vezzi not necessarily. We pull in specific commits from the repo -- as ngi_pipeline doesn't have any releases/tags -- not the latest commit in the master branch. E.g. at the moment it is (will be) `d4563fe` (https://github.com/NationalGenomicsInfrastructure/irma-provision/blob/master/host_vars/127.0.0.1/main.yml#L28). 

Though, of course, if it takes a long time to test this then it might be unwise to merge into the master branch regardless, as it will make it hard for us to get other future changes in to Irma until this has been tested.

**Comment by vezzi** [comment]
@Galithil ... time to merge!!!! surprised!!!! You need to offer @sylvinite a beer...

**Comment by ewels** [comment]
party
